### PR TITLE
Fix graphie processing

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Removes opinionated processing from raw Perseus JSON parsing, and instead treats it as plain text looking for images to replace, in the same way the frontend perseus renderer does
* To still allow the edge case of non-KA perseus JSON with local file references, adds file:// protocol processing to allow references to local file system if needed (adds a test for this)
* Removes any graphie processing from the BaseQuestion and only puts it into the PerseusQuestion type instead
* Updates all related tests, including removing tests for now unused methods. Uses pytest parameterize for better tracing of test failures when looping over fixtures - replace unneeded fixtures with module variables

## References
Needed to fix https://github.com/learningequality/kolibri-library/issues/20

## Reviewer guidance
Have I inadvertently removed some test coverage that might mean that question images are not being properly handled now?

## Summary by CodeRabbit

- **New Features**
  - Added a new interactive Perseus exercise for labeling a bar graph, including stepwise hints and image support.

- **Refactor**
  - Simplified and unified image URL processing in Perseus questions for improved reliability and maintainability.
  - Streamlined test structure by replacing fixtures with parameterized tests and module-level data.
  - Updated `_ExerciseGraphieFile` handling to support language-specific content retrieval and caching.
  - Enhanced question creation to include language context for Perseus exercises.

- **Bug Fixes**
  - Improved handling of local file URLs to ensure compatibility across platforms.

- **Tests**
  - Added tests verifying correct processing of local file URLs with `file://` protocol.
  - Enhanced test coverage for Perseus question processing with parameterized inputs.
  - Updated tests to use parameterized data, increasing clarity and maintainability.
  - Modified tests to support language-aware `_ExerciseGraphieFile` instantiation.

### CODE REVIEW

File: ricecooker/utils/pipeline/transfer.py [hunk 1]
```diff
@@ -41,8 +42,25 @@ class DiskResourceHandler(FileHandler):
...
+
+            # Handle Windows specifics for file:// URLs
+            if platform == "win32":
+                # Remove leading slash for Windows absolute paths
+                if path.startswith("/") and path[1:2].isalpha() and path[2:3] == ":":
+                    path = path[1:]
+                # Replace forward slashes with backslashes for Windows
+                path = path.replace("/", "\\")
+
+        return path
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/utils/pipeline/transfer.py]: Thank you, this was a helpful suggestion. I have updated the code and added tests for the UNC locations.

File: ricecooker/classes/questions.py [hunk 2]
```diff
@@ -307,137 +279,60 @@ def validate(self):
...
+                # These have occasionally been observed in KA graphie URLs.
+                stripped_path = stripped_path[:-2]
+            if stripped_path.startswith("\\n"):
+                # Remove any escaped line break at the start of the URL.
+                # These have occasionally been observed in KA graphie URLs.
+                stripped_path = stripped_path[2:]
+            # Set the proper protocol
+            full_path = path.replace("//", "https://")
+        for file in self.files:
+            if file.path == full_path:
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/questions.py]: Thank you, this was an important catch, I've updated accordingly. You've also helped make the code much more concise.

File: ricecooker/utils/pipeline/transfer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/utils/pipeline/transfer.py]: Hrm, this actually didn't work at all. I've reverted the UNC handling, and instead just kept your suggestions for better normalizing the paths.

File: ricecooker/classes/questions.py [hunk 3]
```diff
@@ -307,137 +279,54 @@ def validate(self):
...
+            exercise_image_file.process_file()
+            self.files.append(exercise_image_file)
+        # Get `new_path` = the replacement path for the image resource
+        new_path = exercises.CONTENT_STORAGE_FORMAT.format(
+            exercise_image_file.get_replacement_str()
+        )
+        if protocol == "web+graphie":  # need to put back the `web+graphie:` prefix
+            new_path = "web+graphie:" + new_path
+        return new_path
+
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/questions.py]: Thanks - I have also reinstated the previous encoded line-break removal to everything, just in case. And reduced the if branching.

File: ricecooker/classes/files.py [hunk 4]
```diff
@@ -429,7 +431,11 @@ def process_file(self):
             config.FAILED_FILES.append(self)
 
     def generate_graphie_file(self):
-        key = "GRAPHIE: {}".format(self.path)
+        if self.ka_language is None:
+            raise ValueError("ka_language must be specified")
+        key = "GRAPHIE: {}".format(
+            self.path + (self.ka_language if self.ka_language != "en" else "")
+        )
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/files.py:439]: Given the fixed length hash generated URLs for KA Graphie content, I don't think this is an actual concern.

File: ricecooker/classes/files.py [hunk 5]
```diff
@@ -396,9 +397,10 @@
...
     default_preset = format_presets.EXERCISE_GRAPHIE
 
-    def __init__(self, path):
+    def __init__(self, path, ka_language):
         self.original_filename = path.split("/")[-1].split(".")[0]
         self.path = path
+        self.ka_language = ka_language
 
     def validate(self):
         try:
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/files.py:406]: This is pre-existing and doesn't seem to be an issue, although I admit, it feels a bit smelly.

File: ricecooker/classes/files.py [hunk 6]
```diff
@@ -447,14 +453,31 @@
...
+        try:
+            r = config.DOWNLOAD_SESSION.get(json_path_base + "-data.json", stream=True)
+            r.raise_for_status()
+        except HTTPError:
+            if self.ka_language == "en":
+                raise
+            r = config.DOWNLOAD_SESSION.get(self.path + "-data.json", stream=True)
+            r.raise_for_status()
+            should_cache = False
         for chunk in r.iter_content():
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/files.py:474]: I only want to try the English version in the case that there's an actual HTTPError in fetching the file though, this is indicative of a 404 or a 400 or something, whereas a timeout suggests it is a transitory issue and should be retried, rather than replaced.

File: ricecooker/classes/questions.py [hunk 7]
```diff
@@ -17,9 +15,7 @@
 from ricecooker.utils.encodings import get_base64_encoding
 
 
-WEB_GRAPHIE_URL_REGEX = (
-    r"web\+graphie:(?P<rawpath>[^\)]+)"  # match web_graphie:{{path}}
-)
+IMAGE_URL_REGEX = r"(?P<protocol>web\+graphie|https?|file|data):(?P<rawpath>[^\)\"]+)"  # match protocol:{{path}}
 MARKDOWN_IMAGE_REGEX = r"!\[([^\]]+)?\]\(([^\)]+?)\)"  # match ![{{smth}}]({{url}})
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/questions.py:19]: Do they? I thought Perseus used markdown. Have you got an example of this?

[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/questions.py:19]: That question is of type `exercises.INPUT_QUESTION` not a Perseus Question as is being handled here. A Perseus Question is represented as raw JSON, and all question text is written in markdown.

File: ricecooker/classes/files.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/files.py:406]: Apparently, this really was an issue. Thanks for some changes in logic in the File base class, this lack of super calling caused a breakage when trying to get the preset.

**Review by marcellamaki — COMMENTED** [review]
Added a couple of questions. (Another perk of the coderabbit review - helps me understand some of the code and decision making. It's nice to come in after that as a reviewer). They're not in and of themselves blocking, but i want to reread this again tomorrow and see if I have any other thoughts or questions, and then I'll go ahead and approve.

File: ricecooker/classes/questions.py [hunk 8]
```diff
@@ -144,10 +140,7 @@ def set_images(self, text, parse_html=True):
         """
         # Set up return values and regex
         file_list = []
-        if parse_html:
-            processed_string = self.parse_html(text)
-        else:
-            processed_string = text
+        processed_string = self.parse_html(text)
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/questions.py:143]: This method used to be used for both !PerseusQuestions and PerseusQuestions. PerseusQuestions should not contain HTML, so we did not parse them as such.

With the changes I made, this method is now only used by !PerseusQuestions so the conditionality was no longer required.

File: ricecooker/classes/files.py [hunk 9]
```diff
@@ -447,14 +453,31 @@ def generate_graphie_file(self):
...
+            json_path_base = self.path
+        else:
+            base_path, _, file_hash = self.path.rpartition("/")
+            json_path_base = base_path + "/" + self.ka_language + "/" + file_hash
+        should_cache = True
+        try:
+            r = config.DOWNLOAD_SESSION.get(json_path_base + "-data.json", stream=True)
+            r.raise_for_status()
+        except HTTPError:
+            if self.ka_language == "en":
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply ricecooker/classes/files.py:470]: The HTTPError indicates there is something wrong with the resource, rather than it just be a temporary connection or timeout issue, so either the resource doesn't exist, is forbidden or whatnot. In that case, we try to get the English version instead, in case it just hasn't been translated yet and that's why we can't access it.

**Review by marcellamaki — APPROVED** [review]
Approving - I cannot promise there are no problems, but I feel at this point like i can follow along well enough with the code and intention, and it's okay to merge
